### ISSUE TITLE
Deleting an object by clicking out from an empty input field is not undoable

### ISSUE THREAD

**Issue by blackears** [musescore/MuseScore#14763]
**Describe the bug**
If you delete all the text in a dynamic marking, it is not possible to undo it.

**To Reproduce**

Create a new score with a piano.
Enter enough notes to fill several measures.
Attack dynamic markings to some of these measures.
Click on one of the dynamic markings and use the backspace key to delete all the letters in it.
Click outside of the text editing area.
Press undo.  The previous text will not be restored.

**Expected behavior**
Pressing undo should restore the text to what it was before the edit.

**Platform information**
Windows 10

**Comment by blackears** [comment]
There is also no way to select the dynamic marking after the text is deleted, so that's also a problem.

**Comment by oktophonie** [comment]
Is the issue here not that _any_ text item (whether dynamics or not), if reduced to empty string and then deselected, simply stops existing? And is that, in fact, a problem, or is it expected behaviour?

**Comment by Tantacrul** [comment]
- I would say it should be un-doable (can't imagine a reason anyone would disagree)
- I think if one deletes **all text** within a text item, it makes sense to just remove the text item. Otherwise, we have to build a new text frame object with new rules. I'm not against doing that in the long-term but not just for fixing this specific issue.

Anyone disagree? If not, I say, let's just fix the undo stack.

**Comment by oktophonie** [comment]
To clarify: deleting all (or any) of the text is undoable, as expected.
But, deleting all the text and then clicking away from (i.e. deselecting) it causes the text item to vanish, and at that point, undo does not recreate it.

Assuming that this is expected (or, at the very least, acceptable) behaviour, then there's nothing to fix and this issue can be closed.

**Comment by blackears** [comment]
Why would that be the expected behavior?  If I accidentally delete the p symbol in a dynamic marking and then click away, I would expect that pressing undo would restore it to as it was before.

**Comment by iainhallam** [comment]
> Why would that be the expected behavior?

Agree with blackears here - if clicking leads to an action then it should be undoable. (For instance, Sibelius even treats selection as an undoable action, so if you've been carefully building up a selection by control-clicking then miss one and accidentally clear the selection, undo restores the previous selection state.)

**Comment by NoahCarverMusic** [comment]
> (For instance, Sibelius even treats selection as an undoable action, so if you've been carefully building up a selection by control-clicking then miss one and accidentally clear the selection, undo restores the previous selection state.)

FYI, the undo selection system that Sibelius uses makes screen reader navigation insanely tedious because one has to go through undo history to find what event they want to undo and what events are simply navigating through notes via the keyboard. I'd much prefer that undo be limited to actions on a score rather than also including selection.

**Comment by davidstephengrant** [comment]
@mathesoncalum Please can you verify that we are not left with empty objects in the project file when clicking out from an empty input field? In this case, the object should be deleted.

Undo should reinstate the object to its state before the edit field was opened (or with the object's default content for newly created objects). Do you agree, @avvvvve? 

https://github.com/user-attachments/assets/bc733250-bf1d-490f-88e5-2b38741727c5

**Comment by mathesoncalum** [comment]
@davidstephengrant confirmed that the empty objects _are_ being deleted in both 4.7 and in master.

For anyone curious this happens in `TextBase::endEdit` (see usage of `newlyAdded`), then later in the `TextEditData` destructor when the edit data is cleared.

**Comment by avvvvve** [comment]
@davidstephengrant Agreed, in that video, clicking undo should restore the text box and it should say 'test'
